subway v0.0.1

shridharama opened the pull request:
Main files:
1. BaseSubwayJob.scala
2. MySqlToSnowflakeSubwayJob.scala
3. MySqlUtils.scala

Apart from code clarity, some high level things that I would appreciate getting a review on:
1. Ease of use of subway from an end-user point of view (can be viewed along with the corresponding configuration PR)
2. Any general scala-service best practices that I might have violated. A lot of the boilerplate was copied from the scala-service template + `surtch`'s backfill module cc @liamstewart

liamstewart commented on project/Build.scala:
```diff
@@ -0,0 +1,23 @@
+
+object Build extends CommonsBuild {
+  val testSetting = libraryDependencies ++= Seq(
+    "org.scalatest" %% "scalatest" % "2.1.7" % "test",
```
There are helpers for these - `scalaTest` and `scalaMock`.

liamstewart commented on project/Build.scala:
```diff
@@ -0,0 +1,23 @@
+        libraryDependencies ++= Seq(
+          stravaCommons("sql"),
+          stravaCommons("spark"),
+          "com.github.pureconfig" %% "pureconfig" % "0.7.2"
```
There is a helper for this `pureConfig`.

liamstewart commented on subway-core/src/main/scala/com/strava/subway/backfill/MySqlToSnowflakeSubwayJob.scala:
```diff
@@ -0,0 +1,138 @@
+
+  /* TODO: Evaluate whether or not we need to cast the DataFrame to a DataSet
+  def readDataset(segmentEffortsTable: DataFrame): DataSet = {
+    segmentEffortsTable
```
Remove all of this commented out code?

liamstewart commented on subway-core/src/main/scala/com/strava/subway/backfill/MySqlToSnowflakeSubwayJob.scala:
```diff
@@ -0,0 +1,138 @@
+  def transform(df: DataFrame) : DataFrame = {
+
+    val repartitionedDf = if(jobConfig.getBoolean("transform.repartition")){
+      df.repartition(16)
```
Any particular reason why this is hard-coded?

liamstewart commented on subway-core/src/main/scala/com/strava/subway/backfill/MySqlToSnowflakeSubwayJob.scala:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    val s3Df = if(jobConfig.getBoolean("transform.writeIntermediateToS3")){
+      val s3Path = s"s3a://strava.scratch/shridhar-tmp/${sourceTableConfig.name}"
```
Path needs to be configurable.

liamstewart commented on subway-core/src/main/scala/com/strava/subway/backfill/MySqlToSnowflakeSubwayJob.scala:
```diff
@@ -0,0 +1,138 @@
+
+import org.apache.spark.sql.{DataFrame, Dataset}
+
+object MySqlToSnowflakeSubwayJob extends BaseSubwayJob[SegmentEffort] {
```
This is supposed to be a general job, right? If so, why the hardcoding to `SegmentEffort`?

liamstewart commented on subway-core/src/main/scala/com/strava/subway/backfill/BaseSubwayJob.scala:
```diff
@@ -0,0 +1,100 @@
+                        columns: Option[List[String]]
+                      )
+
+trait BaseSubwayJob[D] extends Job {
```
What is the `D` type parameter for?

liamstewart commented on README.md:
```diff
@@ -1,2 +1,65 @@
+#### Usage:
+(This will be updated in greater detail in due time)
+For the moment, please check the following sample configs:
+Refresh job: configuration/services/subway/subway-lap-efforts-refresh-prod.conf
```
Use links so I don't have to copy/paste.

shridharama commented:
@liamstewart I have addressed most of your comments. Couple of things that I'm still working on:

1. Unbundling the source and sink to different traits - this is a great suggestion and I'm working on an elegant way to do this. The specific implementation will depend a bit on the next point though.


2. Regarding the source and sink configurations - the issue here is that different data sources/sinks have different standards for the knobs that I am exposing. For ex: when it comes to `parallelism`, both reads and writes to MySQL have a `numPartitions` flag that determines parallelism. I did a quick audit of all performance related configuration parameters for different data sources here - https://docs.google.com/document/d/1HZhJVesXIdAS0Yrgy9UXj-E6LegAiyB7exfO3xK7CRU/edit#

We can probably have a better discussion offline, but based on the above doc, here are some possible approaches that I've thought of:
 - Expose all these knobs (`parallelism`, `interval`, etc) as optional. Document the specific configurations that are required for reading from and writing to different sources and sinks. Have subway validate the user provided configs to ensure that it has all the information it needs for the particular source and sink that it's operating on.

 - Have different `table_config`s for different data sources. So, `mysql_table_config` and `snowflake_table_config` will have different configurations specific to it's spark configurations. Might need to further split it into `source` and `sink`.

I also think `interval` is more a property of subway than of `source`/`sink`, so we should move it to under `strava.subway`.
